spec: Move FLP polynomials to Lagrange basis. (#601)
cfrg/draft-irtf-cfrg-vdaf: draft-irtf-cfrg-vdaf.md

    date: 2025
    target: https://mailarchive.ietf.org/arch/msg/cfrg/Omdhr4rO1pla_nlju2l7OJEGWPM/

>   Pol71: DOI.10.1090/S0025-5718-1971-0301966-0

Reviewer: That's less important in my opinion than providing a useful reference. This is a spec afterall :)

The following function denotes polynomial evaluation in the
coefficient representation.

- `poly_eval(field: type[F], p: list[F], x: F) -> F` returns the
  evaluation of the polynomial `p` (in the monomial basis) at `x`.
  The `field` parameter is the class object for `F` and is used by
  the implementation to construct field elements. (See {{field}}.)

#### Lagrange Basis

In the Lagrange basis, polynomials are represented as a list of
evaluations over a set of points.
This set is fixed to the first `n` powers of the
principal `n`-th root of unity.
So any polynomial `p` of degree lesser than `n` is represented as a
list `v` of `n` values such that `v[i] = p(Wn ** i)`,
where `Wn = Field.nth_root(n)`.

The `Lagrange` class defines operations over polynomials in the
Lagrange basis. It implements the following methods:

- `poly_mul(self, p: list[F], q: list[F]) -> list[F]` multiplies two
  polynomials, where the input and output polynomials are in the Lagrange basis.
  The length of inputs must be equal and be a power of two.

- `poly_eval(self, p: list[F], x: F) -> F` evaluates a polynomial
  `p` (in the Lagrange basis) at `x` in linear time.

>   Avoids conversion to the monomial basis as shown below.

Reviewer 2: ```suggestion
  It avoids conversion to the monomial basis as shown below.
```

### NTT-Friendly Fields {#field-ntt-friendly}

Some VDAFs, including Prio3, require fields that are suitable for efficient
computation of the number theoretic transform (NTT) {{Pol71}}, as this allows
for fast conversion between polynomial representations. Refer to {{SML24}}
for an overview regarding NTT.

> Specifically, a field is said to be "NTT-friendly" if, in addition to the

Reviewer: I think we need to explain the difference between Lagrange and monomial basis, as these terms are used in the remainder assuming the reader knows what they mean. How about this:
```suggestion

The NTT operation above is used to convert a polynomial from its representation in the monomial basis to its representation in the Lagrange basis. The monomial basis is the standard way of representing a polynomial, where each element of the list corresponds to one of the coefficients of the polynomial. In the Lagrange basis, each element corresponds to the evaluation of a polynomial at a special value known as a root of unity. This representation allows for certain computations to be carried out more efficiently than in the monomial basis. See {{Faz25}} for details.

A field is said to be "NTT-friendly" if, in addition to the
```

  `Field.gen()`. This is the smallest positive integer for which
  `Field.gen() ** Field.GEN_ORDER == Field(1)`.

* `Field.nth_root_powers(n: int) -> list[F]` returns the first `n` powers of
  an `n`-th root of unity.

* `Field.ntt(p: list[F], n: int, set_s: bool = False) -> list[F]` returns

>   the values `p(w**i)` for `0 ≤ i < n`, where `p` are the coefficients a

Reviewer: ```suggestion
  the values `p(w**i)` for `0 ≤ i < n`, where `p` are the coefficients of a
```

  polynomial, `w` is an `n`-th root of unity, and `n` must be a power of two.
  By passing `set_s=True`, the function returns `p(s * w**i)` for `0 ≤ i < n`,

>   where `s` is a `2n`-th root of unity.

Reviewer: Which root of unity? We need to specify it. (Similar comment above.)

* `Field.inv_ntt(v: list[F], n: int) -> list[F]` returns the coefficients of
  a polynomial `p` such that `v` are the evaluations of `p` at the first `n`

>   powers of an `n`-th root of unity. Note that `n` must be a power of two.

Reviewer: Which root of unity?

The size of the subgroup dictates how large the polynomials can be. It
is RECOMMENDED that a generator is chosen with order at least `2**20`.

The following functions are used to operate with polynomials (over an
NTT-friendly field) represented in the Lagrange basis.

> - `poly_mul(self, p: list[T], q: list[T]) -> list[T]` multiplies

Reviewer: Use `F` instead of `T` for consistency with the remainder of the document. Likewise for the reference implementation.

  two polynomials such that the input and output are in the Lagrange basis

>   representation.

Reviewer: The lengths of the polynomials must be equal and be a power of 2.

  `Field.gen()`. This is the smallest positive integer for which
  `Field.gen() ** Field.GEN_ORDER == Field(1)`.

* `Field.nth_root_powers(n: int) -> list[F]` returns the first n powers of
  an `n`-th root of unity.

> 

Reviewer: We need to add `ntt()` here don't we?

    # Construct the proof, which consists of the wire seeds and
    # gadget polynomial for each gadget.
    proof = []
    for g, g_calls in zip(valid.GADGETS, valid.GADGET_CALLS):
        g = cast(ProveGadget[F], g)

        # Define `p` as the smallest power of two accommodating all
        # gadget calls plus one.

>         p = next_power_of_2(1 + g_calls)

Author: done.

        # The validity circuit evaluation defines one polynomial for
        # each input wire of each gadget.
        # For each wire `j`, the vector `g.wires[j]` of length `p`
        # is built as follows:
        # - `g.wires[j][0]` is set to the seed for wire `j` (from
        #   the prover's randomness).
        # - The subsequent entries are the assigned values from each
        #   gadget call.
        # - Pad the vector with zeros to reach length `p`.
        # The wire polynomial is then defined by its evaluations:
        #   `wire_poly(alpha**k) = g.wires[j][k]`
        # for all `k`, where `alpha` is a `p`-th root of unity.
        wire_polys = [g.wires[j] for j in range(g.ARITY)]
        wire_seeds = [g.wires[j][0] for j in range(g.ARITY)]
        proof += wire_seeds

        # Compute the gadget polynomial by evaluating the gadget
        # on the wire polynomials. By construction we have that
        # `gadget_poly(alpha**k)` is the `k`-th output.
        gadget_poly = g.eval_poly(self.field, wire_polys)
        gadget_poly_len = g.DEGREE * (p - 1) + 1

>         proof += gadget_poly[:gadget_poly_len]

Author: done.

also evaluates the gadget polynomial for each gadget invocation in order to
produce the gadget's output. Then compute the wire polynomials from the
recorded values.
Note that these polynomials are in the value (Lagrange) representation

> by construction.

Reviewer 2: Here's a suggested rephrasing, to reflect that only the wire polynomials are constructed during the query process, while the gadget polynomials are received from the client.
```suggestion
Note that the wire polynomials are in the value (Lagrange) representation
by construction, and the gadget polynomials parsed from the proof share
are similarly transmitted in value (Lagrange) representation.
```

Next, choose a random point `t` (parsed from the query randomness), evaluate
each wire polynomial at `t`, and evaluate the gadget polynomial at `t`.
Polynomial evaluation in the value (Lagrange) representation is
faster by using the algorithm by {{Faz25}} without resorting to
polynomial interpolation.
The results are recorded in the verifier message passed to the decision

> algorithm, where the test is finished.

Reviewer 2: In the next paragraph, we should probably change "The random point `t` MUST NOT be one of the fixed evaluation points used to interpolate the wire polynomials." to "... used to define the wire polynomials."
